src/FileConventions/Library.fs

```diff
@@ -104,3 +103,78 @@ let EolAtEof(fileInfo: FileInfo) =
                 False
         else
             True
+
+let DetectInconsistentVersion
+    (fileInfos: seq<FileInfo>)
+    (versionRegexPattern: string)
+    =
+    let versionRegex = Regex(versionRegexPattern, RegexOptions.Compiled)
+
+    let mutable versionMap: Map<string, Set<string>> = Map.empty
+
+    fileInfos
+    |> Seq.iter(fun fileInfo ->
+        let fileText = File.ReadAllText fileInfo.FullName
+
+        versionRegex.Matches fileText
+        |> Seq.iter(fun regexMatch ->
+            let key = regexMatch.Groups.[1].ToString()
+            let value = regexMatch.Groups.[2].ToString()
+
+            let addSet (value: string) (maybePrevSet: Option<Set<string>>) =
+                match maybePrevSet with
+                | Some prevSet -> Some(Set.add value prevSet)
+                | None -> None
+
+            if versionMap.ContainsKey key then
+                versionMap <- versionMap.Change(key, addSet value)
+            else
+                versionMap <- versionMap.Add(key, Set.singleton value)
+        )
+    )
+
+    versionMap
+    |> Seq.map(fun item -> Seq.length item.Value > 1)
+    |> Seq.contains true
+
+let DetectInconsistentVersionsInGitHubCIWorkflow(fileInfos: seq<FileInfo>) =
+    fileInfos
+    |> Seq.iter(fun fileInfo -> assert (fileInfo.FullName.EndsWith ".yml"))
+
+    let inconsistentVersionsType1 =
+        DetectInconsistentVersion
+            fileInfos
+            "\\swith:\\s*([^\\s]*)-version:\\s*([^\\s]*)\\s"
+
+    let inconsistentVersionsType2 =
+        DetectInconsistentVersion
+            fileInfos
+            "\\suses:\\s*([^\\s]*)@v([^\\s]*)\\s"
+
+    inconsistentVersionsType1 || inconsistentVersionsType2
+
+let DetectInconsistentVersionsInGitHubCI(dir: DirectoryInfo) =
+    let ymlFiles = dir.GetFiles("*.yml", SearchOption.AllDirectories)
+
+    if Seq.length ymlFiles = 0 then
+        false
+    else
+        DetectInconsistentVersionsInGitHubCIWorkflow ymlFiles
+
+let DetectInconsistentVersionsInNugetReferencesInFsharpScripts
+    (fileInfos: seq<FileInfo>)
+    =
+    fileInfos
+    |> Seq.iter(fun fileInfo -> assert (fileInfo.FullName.EndsWith ".fsx"))
+
+    DetectInconsistentVersion
+        fileInfos
+        "#r \"nuget:\\s*([^\\s]*)\\s*,\\s*Version\\s*=\\s*([^\\s]*)\\s*\""
+
+let DetectInconsistentVersionsInFsharpScripts(dir: DirectoryInfo) =
+    let fsxFiles = dir.GetFiles("*.fsx", SearchOption.AllDirectories)
+
+    if Seq.length fsxFiles = 0 then
+        false
+    else
+        DetectInconsistentVersionsInNugetReferencesInFsharpScripts fsxFiles
```

Comment:
@realmarv nit: let's use FSharp, not Fsharp; also, use Refs instead of References, to make this func name not so long

Reply:
Done